### DISCUSSION TITLE
Mesh output?

### DISCUSSION THREAD

**Discussion by MattiaCasalegno-PRATT** [NVlabs/instant-ngp#412] [Q&A]
Hello, thank you for sharing this amazing project.
I was wondering what types of meshes can we export? And are those textured or vertex colored? I'm interested in exporting 3d reconstructed scenes into Blander or meshlab.
Thank you

**Comment by Angkitbharata** [comment]
> Hello, thank you for sharing this amazing project.
> I was wondering what types of meshes can we export? And are those textured or vertex colored? I'm interested in exporting 3d reconstructed scenes into Blander or meshlab.
> Thank you

**Comment by MarvinChung** [comment]
There is a save_mesh argument in the scripts [scripts/run.py](https://github.com/NVlabs/instant-ngp/blob/master/scripts/run.py) which supports OBJ and PLY format.   

First prepare a folder to save the model and mesh.
```
mkdir saved  
```
Then run the script 
```
python3 ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --gui --train
```
You can get the fox.ply in the saved directory and then export it to blender. It takes a while for me to export it to the blender.  
My result looks like this:
![image](https://user-images.githubusercontent.com/22186310/160994164-fd479cfd-dfc8-4cf3-ad01-1b1f721589d8.png)

**Reply by entropical** [reply]
Hi Marvin, thanks for this answer. Do you know how you would be able to include colour information in this? Ultimately I would like to convert it to a pointcloud, and render it with Potree. Some samples in Potree include colour information.

**Reply by rozeappletree** [reply]
> There is a save_mesh argument in the scripts [scripts/run.py](https://github.com/NVlabs/instant-ngp/blob/master/scripts/run.py) which supports OBJ and PLY format.
> 
> First prepare a folder to save the model and mesh.
> 
> ```
> mkdir saved  
> ```
> 
> Then run the script
> 
> ```
> python3 ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --gui --train
> ```
> 
> You can get the fox.ply in the saved directory and then export it to blender. It takes a while for me to export it to the blender. My result looks like this: ![image](https://user-images.githubusercontent.com/22186310/160994164-fd479cfd-dfc8-4cf3-ad01-1b1f721589d8.png)

@Tom94 haven't you mentioned exporting is not possible with current codebase? What exactly is the difference here with -save_mesh flag

**Comment by aceIsHere** [comment]
Hey man what will be the equivalent like for google colab for this notebook (https://colab.research.google.com/drive/10TgQ4gyVejlHiinrmm5XOvQQmgVziK3i?usp=sharing)

Basically after last line I want to export/save my file as a .obj or .ply format so that I can download and try it on my local blender environment

I tried to iterate your line for colab but it was not successful. I hope you can help me here. Thanks!

**Reply by aceIsHere** [reply]
If possible can you point out the source where I can now about the relation 3d objects with the amount of iterations (in general the more the better but a clearer understanding is needed)

**Reply by MarvinChung** [reply]
Can you show the error on colab?  

Did you remove the --gui option in your colab?  
```
python3 ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --train
```

**Reply by aceIsHere** [reply]
The whole notebooks runs smoothly till the last black i.e.

!./build/testbed --scene data/nerf/fox

After that just when I put the above command (ur command to export) it simply prints invalid syntax

**Reply by MarvinChung** [reply]
Did you add the exclamation mark in the colab?  
```
!mkdir saved
!python3 ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --train
```

**Reply by aceIsHere** [reply]
Will try ASAP and will msg back

**Reply by aceIsHere** [reply]
Used the "!" now and everything went successfully up until about the ending and then the following error log was printed

13:47:49 INFO     Loading NeRF dataset from
13:47:49 INFO       ./data/nerf/fox/transforms.json
13:47:49 PROGRESS []   0% ( 0/50)  0s/inf13:47:49 PROGRESS []   2% ( 1/50) 0s/1s13:47:49 PROGRESS []   4% ( 2/50) 0s/0s13:47:49 PROGRESS []   6% ( 3/50) 0s/0s13:47:49 PROGRESS []   8% ( 4/50) 0s/0s13:47:49 PROGRESS []  10% ( 5/50) 0s/0s13:47:49 PROGRESS []  12% ( 6/50) 0s/0s13:47:49 PROGRESS []  14% ( 7/50) 0s/0s13:47:49 PROGRESS []  16% ( 8/50) 0s/0s13:47:49 PROGRESS []  18% ( 9/50) 0s/0s13:47:49 PROGRESS []  20% (10/50) 0s/0s13:47:49 PROGRESS []  22% (11/50) 0s/0s13:47:49 PROGRESS []  24% (12/50) 0s/0s13:47:49 PROGRESS []  26% (13/50) 0s/0s13:47:49 PROGRESS []  28% (14/50) 0s/0s13:47:49 PROGRESS []  30% (15/50) 0s/0s13:47:49 PROGRESS []  32% (16/50) 0s/0s13:47:49 PROGRESS []  34% (17/50) 0s/0s13:47:49 PROGRESS []  36% (18/50) 0s/0s13:47:49 PROGRESS []  38% (19/50) 0s/0s13:47:49 PROGRESS []  40% (20/50) 0s/0s13:47:49 PROGRESS []  42% (21/50) 0s/0s13:47:49 PROGRESS []  44% (22/50) 0s/0s13:47:49 PROGRESS []  46% (23/50) 0s/0s13:47:49 PROGRESS []  48% (24/50) 0s/0s13:47:49 PROGRESS []  50% (25/50) 0s/0s13:47:49 PROGRESS []  52% (26/50) 0s/0s13:47:49 PROGRESS []  54% (27/50) 0s/0s13:47:49 PROGRESS []  56% (28/50) 0s/0s13:47:49 PROGRESS []  58% (29/50) 0s/0s13:47:49 PROGRESS []  60% (30/50) 0s/0s13:47:49 PROGRESS []  62% (31/50) 0s/0s13:47:49 PROGRESS []  64% (32/50) 0s/0s13:47:49 PROGRESS []  66% (33/50) 0s/0s13:47:49 PROGRESS []  68% (34/50) 0s/0s13:47:49 PROGRESS []  70% (35/50) 0s/0s13:47:49 PROGRESS []  72% (36/50) 0s/0s13:47:49 PROGRESS []  74% (37/50) 0s/0s13:47:49 PROGRESS []  76% (38/50) 0s/0s13:47:49 PROGRESS []  78% (39/50) 0s/0s13:47:49 PROGRESS []  80% (40/50) 0s/0s13:47:49 PROGRESS []  82% (41/50) 0s/0s13:47:49 PROGRESS []  84% (42/50) 0s/0s13:47:49 PROGRESS []  86% (43/50) 0s/0s13:47:49 PROGRESS []  88% (44/50) 0s/0s13:47:49 PROGRESS []  90% (45/50) 0s/0s13:47:49 PROGRESS []  92% (46/50) 0s/0s13:47:49 PROGRESS []  94% (47/50) 0s/0s13:47:49 PROGRESS []  96% (48/50) 0s/0s13:47:49 PROGRESS []  98% (49/50) 0s/0s13:47:49 PROGRESS [] 100% (50/50) 0s/0s13:47:49 SUCCESS  Loaded 50 images of size 1080x1920 after 0s
13:47:49 INFO       cam_aabb=[min=[1.0229,-1.33309,-0.378748], max=[2.46175,1.00721,1.41295]]
13:47:50 INFO     Loading network config from: /content/instant-ngp/configs/nerf/base.json
13:47:50 INFO     GridEncoding:  Nmin=16 b=1.51572 F=2 T=2^19 L=16
13:47:50 INFO     Density model: 3--[HashGrid]-->32--[FullyFusedMLP(neurons=64,layers=3)]-->1
13:47:50 INFO     Color model:   3--[Composite]-->16+16--[FullyFusedMLP(neurons=64,layers=4)]-->3
13:47:50 INFO       total_encoding_params=13074912 total_network_params=10240
Training: 100% 99984/100000 [39:04<00:00, 42.64step/s, loss=0.000274]
Saving snapshot  saved/fox.msgpack
Generating mesh via marching cubes and saving to saved/fox.ply. Resolution=[256,256,256]
14:26:55 INFO     #vertices=3503145 #triangles=6865914
Traceback (most recent call last):
  File "./scripts/run.py", line 295, in <module>
    testbed.compute_and_save_marching_cubes_mesh(args.save_mesh, [res, res, res])
RuntimeError: Failed to open saved/fox.ply for writing.

**Reply by MarvinChung** [reply]
Sorry, I have no idea. Maybe there is not enough disk space?

**Reply by aceIsHere** [reply]
Trying to figure out what line 295 is all about in the scripts/run.py, But still thanks for helping me out :)

**Reply by aceIsHere** [reply]
Issue solved refer to https://github.com/NVlabs/instant-ngp/issues/427

**Reply by MarvinChung** [reply]
I see. I  will also edit my solution to prevent confusion.

**Reply by LincolnG17** [reply]
I can't use this command either. I have tried both my own nerf and the default fox:
(base) PS C:\ngp\instant-ngp> python ./scripts/run.py --mode nerf --scene ./data/nerf/fox/ --save_snapshot saved/fox.msgpack --save_mesh saved/fox.ply --gui --train
Warning: the '--mode' argument is no longer in use. It has no effect. The mode is automatically chosen based on the scene.
12:28:12 SUCCESS  Initialized CUDA. Active GPU is #0: NVIDIA GeForce RTX 3060 Ti [86]
12:28:12 INFO     Loading NeRF dataset from
12:28:12 INFO       .\data\nerf\fox\transforms.json
12:28:12 SUCCESS  Loaded 50 images after 0s
12:28:12 INFO       cam_aabb=[min=[1.0229,-1.33309,-0.378748], max=[2.46175,1.00721,1.41295]]
12:28:14 SUCCESS  Initialized Vulkan and NGX on GPU #0: NVIDIA GeForce RTX 3060 Ti
Training:   0%|                                                                            | 0/35000 [00:00<?, ?step/s]12:28:14 INFO     Loading network config from: .\configs\nerf\base.json
12:28:14 INFO     GridEncoding:  Nmin=16 b=1.51572 F=2 T=2^19 L=16
12:28:14 INFO     Density model: 3--[HashGrid]-->32--[FullyFusedMLP(neurons=64,layers=3)]-->1
12:28:14 INFO     Color model:   3--[Composite]-->16+16--[FullyFusedMLP(neurons=64,layers=4)]-->3
12:28:14 INFO       total_encoding_params=13074912 total_network_params=10240
Training:  27%|█████████████▋                                    | 9557/35000 [01:57<04:57, 85.45step/s, loss=0.000416]12:30:12 WARNING  Vulkan error: Validation Error: [ VUID-VkImageCreateInfo-extent-00944 ] Object 0: handle = 0x1e6baecb790, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x8869da08 | vkCreateImage: parameter pCreateInfo->extent.width (= 0) is not greater than 0 The Vulkan spec states: extent.width must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-VkImageCreateInfo-extent-00944)
12:30:12 WARNING  Vulkan error: Validation Error: [ VUID-VkImageCreateInfo-extent-00945 ] Object 0: handle = 0x1e6baecb790, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xeaa98d2c | vkCreateImage: parameter pCreateInfo->extent.height (= 0) is not greater than 0 The Vulkan spec states: extent.height must be greater than 0 (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-VkImageCreateInfo-extent-00945)
Training:  27%|█████████████▋                                    | 9557/35000 [01:57<05:12, 81.44step/s, loss=0.000416]
Traceback (most recent call last):
  File "C:\ngp\instant-ngp\scripts\run.py", line 179, in <module>
    while testbed.frame():
RuntimeError: C:\ngp\instant-ngp\src\dlss.cu:657 cudaGetMipmappedArrayLevel(&first_level_array, m_cuda_mipmapped_array, 0) failed with error invalid resource handle
12:30:12 WARNING  Vulkan error: Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x1e6baecb790, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xa9fdf000000014c, type = VK_OBJECT_TYPE_IMAGE; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x1e6baecb790[], VkImage 0xa9fdf000000014c[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)
12:30:12 WARNING  Vulkan error: Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x1e6baecb790, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xe8fb46000000014d, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x1e6baecb790[], VkDeviceMemory 0xe8fb46000000014d[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)
12:30:12 WARNING  Vulkan error: Validation Error: [ VUID-vkDestroyDevice-device-00378 ] Object 0: handle = 0x1e6baecb790, type = VK_OBJECT_TYPE_DEVICE; Object 1: handle = 0xd406ad000000014e, type = VK_OBJECT_TYPE_IMAGE_VIEW; | MessageID = 0x71500fba | OBJ ERROR : For VkDevice 0x1e6baecb790[], VkImageView 0xd406ad000000014e[] has not been destroyed. The Vulkan spec states: All child objects created on device must have been destroyed prior to destroying device (https://vulkan.lunarg.com/doc/view/1.3.239.0/windows/1.3-extensions/vkspec.html#VUID-vkDestroyDevice-device-00378)

**Comment by Angkitbharata** [comment]
So thankyou

Pada tanggal Sel, 5 Apr 2022 01.28, aceIsHere ***@***.***>
menulis:

**Reply by barakmam** [reply]
For the color you can use MeshLab software

**Comment by tempdeltavalue** [comment]
Hi, how to add color inside blender (not Meshlab) ?

**Reply by lucasew** [reply]
Just use meshlab, its way easier and its free

**Reply by tempdeltavalue** [reply]
try to launch camera view , blender doesn't render color in "object/edit" mode 🤷‍♂️ (F12)
